File: backend/models/Task.js
// backend/models/Task.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

/**
 * Reusable string array sanitizer
 */
function normalizeStringArray(value) {
  if (!Array.isArray(value)) return [];
  return value
    .map((item) => String(item || '').trim())
    .filter(Boolean);
}

/**
 * Submission Sub Schema
 * Tracks multiple pieces of work uploaded by the student.
 */
const submissionSchema = new Schema(
  {
    student: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'Submission student is required'],
    },
    
    // ============================================================
    // MODIFICATION: MULTI-FILE SUPPORT
    // ============================================================
    files: [
      {
        url: { 
          type: String, 
          required: [true, 'File URL is required'],
          trim: true 
        },
        name: { 
          type: String, 
          required: [true, 'File name is required'],
          trim: true 
        }
      }
    ],

    // BACKWARDS COMPATIBILITY: Keeps old single-file tasks working
    fileUrl: {
      type: String,
      required: false,
      trim: true
    },

    notes: {
      type: String,
      default: '',
      trim: true,
      maxlength: [2000, 'Submission notes cannot exceed 2000 characters'],
    },
    
    approved: {
      type: Boolean,
      default: false,
    },
    submittedAt: {
      type: Date,
      default: Date.now,
    },
    clientApprovedAt: {
      type: Date,
      default: null,
    }
  },
  { _id: false }
);

/**
 * Main Task Schema
 */
const taskSchema = new Schema(
  {
    /**
     * Basic Info
     */
    title: {
      type: String,
      required: [true, 'Task title is required'],
      trim: true,
      maxlength: [150, 'Task title cannot exceed 150 characters'],
    },

    description: {
      type: String,
      required: [true, 'Task description is required'],
      trim: true,
      maxlength: [5000, 'Task description cannot exceed 5000 characters'],
    },

    /**
     * CLIENT ATTACHMENTS (PROJECT BRIEF)
     * Files provided by the client when creating the task.
     */
    attachments: {
      type: [String],
      default: [],
    },
    attachmentNames: {
      type: [String],
      default: [],
    },

    /**
     * Client Logic: Registered vs Guest (Emergency Task)
     */
    isGuestTask: {
      type: Boolean,
      default: false,
    },

    client: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: false,
      index: true,
    },

    guestInfo: {
      name: { type: String, trim: true },
      mobile: { type: String, trim: true },
      email: { type: String, trim: true },
    },

    /**
     * Workflow Statuses
     */
    status: {
      type: String,
      enum: [
        'open', 
        'request_sent', 
        'assigned', 
        'under_review', 
        'completed', 
        'declined'
      ],
      default: 'open',
      index: true,
    },

    /**
     * Gated Access Control
     */
    clientCanViewSubmission: {
      type: Boolean,
      default: true,
    },

    clientCanDownload: { 
      type: Boolean, 
      default: false 
    },

    /**
     * Manual Payment Chain Tracking
     */
    adminReceivedPayment: {
      type: Boolean,
      default: false,
    },

    adminPaidStudent: {
      type: Boolean,
      default: false,
    },

    budgetFinalized: {
      type: Boolean,
      default: false
    },

    /**
     * Project Parameters
     */
    
    // ============================================================
    // MODIFICATION: DUAL FINANCIAL FIELDS
    // ============================================================
    // budget: What the CLIENT pays the platform
    budget: {
      type: Number,
      required: false, 
      min: [0, 'Budget cannot be negative'],
    },
    
    // studentPayout: What the STUDENT actually receives (Hidden from Client)
    studentPayout: {
      type: Number,
      required: false,
      min: [0, 'Payout cannot be negative'],
      default: 0
    },
    // ============================================================

    deadline: {
      type: Date,
      required: [true, 'Deadline is required'],
    },

    location: {
      type: String,
      trim: true,
      default: '',
    },

    domain: {
      type: String,
      trim: true,
      default: '',
      index: true,
    },

    requiredSkills: {
      type: [String],
      default: [],
      set: normalizeStringArray,
    },

    /**
     * Matching & Accountability
     */
    assignedByAdmin: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      default: null,
    },

    assignedAt: {
      type: Date,
      default: null,
    },

    student: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      default: null,
      index: true,
    },

    requestedStudent: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      default: null,
    },

    assignmentRequestStatus: {
      type: String,
      enum: [null, 'request_sent', 'request_rejected'],
      default: null,
    },

    /**
     * Deliverables & Feedback
     */
    submission: {
      type: submissionSchema,
      default: null,
    },

    modificationNotes: {
      type: String,
      default: '',
      trim: true
    },

    attemptCount: { type: Number, default: 0 },

    rating: { type: Number, default: 0 },
    feedback: { type: String, default: '', trim: true },
    score: { type: Number, default: 0 },
  },
  {
    timestamps: true,
    versionKey: false,
  }
);

/**
 * Pre-validation cleanup
 */
taskSchema.pre('validate', function () {
  this.title = String(this.title || '').trim();
  this.description = String(this.description || '').trim();
  this.location = String(this.location || '').trim();
  this.domain = String(this.domain || '').trim();

  if (this.isGuestTask && this.guestInfo) {
    if (this.guestInfo.name) this.guestInfo.name = String(this.guestInfo.name).trim();
    if (this.guestInfo.mobile) this.guestInfo.mobile = String(this.guestInfo.mobile).trim();
  }
});

/**
 * Business rule enforcement
 */
taskSchema.pre('validate', function () {
  if (!this.isGuestTask && !this.client) {
    throw new Error('Registered tasks require a client reference');
  }

  if (this.isGuestTask && (!this.guestInfo || !this.guestInfo.name || !this.guestInfo.mobile)) {
    throw new Error('Emergency tasks require name and mobile number');
  }

  if (['assigned', 'under_review', 'completed'].includes(this.status)) {
    if (!this.student) {
      throw new Error('Assigned student is required for active or completed projects');
    }
  }
});

module.exports = mongoose.model('Task', taskSchema);

File: frontend/lib/models/task.dart
// lib/models/task.dart

class Task {
  final String id;
  final String title;
  final String description;

  final String? clientId; // Optional for guest tasks
  final String? clientName;

  final double? budget; // Requirement: Estimated amount is optional
  final String status;

  final String? location;
  final String? domain;
  final String? company;
  final List<String> requiredSkills;
  final String? deadline;

  final List<String> attachments;
  final List<String> attachmentNames;

  final String? requestedStudentId;
  final String? requestedStudentName;

  final String? assignedStudentId;
  final String? assignedStudentName;

  final bool clientTermsAccepted;
  final bool studentTermsAccepted;

  // ============================================================
  // MODIFICATION: MULTI-FILE SUBMISSION SUPPORT
  // Replaced single submissionFile string with a list of file objects
  // Each map contains {'url': '...', 'name': '...'}
  // ============================================================
  final List<Map<String, String>> submissionFiles; 
  final String? submissionNotes;
  final String? submittedAt;

  final bool submissionApproved;
  final bool clientApproved;

  final int rating;
  final int score;
  final String feedback;

  final bool isGuestTask;
  final String? guestName;
  final String? guestMobile;
  final String? guestEmail;

  // --- GATING LOGIC PROPERTIES ---
  final bool clientCanViewSubmission;
  final bool clientCanDownload;

  // --- PAYMENT CHAIN TRACKING ---
  final bool adminReceivedPayment;
  final bool adminPaidStudent;

  // --- HYBRID PAYMENT SWITCH ---
  final bool budgetFinalized;

  // Stores the reason/notes provided when a client requests modification.
  final String? modificationNotes;

  const Task({
    required this.id,
    required this.title,
    required this.description,
    this.clientId,
    this.clientName,
    this.budget,
    required this.status,
    this.location,
    this.domain,
    this.company,
    this.requiredSkills = const [],
    this.deadline,
    this.attachments = const [],
    this.attachmentNames = const [],
    this.requestedStudentId,
    this.requestedStudentName,
    this.assignedStudentId,
    this.assignedStudentName,
    this.clientTermsAccepted = false,
    this.studentTermsAccepted = false,
    this.submissionFiles = const [], // Default to empty list
    this.submissionNotes,
    this.submittedAt,
    this.submissionApproved = false,
    this.clientApproved = false,
    this.rating = 0,
    this.score = 0,
    this.feedback = '',
    this.isGuestTask = false,
    this.guestName,
    this.guestMobile,
    this.guestEmail,
    this.clientCanViewSubmission = true,
    this.clientCanDownload = false,
    this.adminReceivedPayment = false,
    this.adminPaidStudent = false,
    this.budgetFinalized = false,
    this.modificationNotes = '', 
  });

  static double? _toDoubleOrNull(dynamic value) {
    if (value == null) return null;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value);
    return null;
  }

  static bool _toBool(dynamic value, {bool fallback = false}) {
    if (value is bool) return value;
    if (value is String) {
      final v = value.trim().toLowerCase();
      return v == 'true' || v == '1';
    }
    if (value is num) return value != 0;
    return fallback;
  }

  static List<String> _toStringList(dynamic value) {
    if (value is List) {
      return value
          .map((e) => e?.toString() ?? '')
          .where((e) => e.trim().isNotEmpty)
          .toList();
    }
    return const [];
  }

  // Helper to parse the new files array in submission
  static List<Map<String, String>> _toFileMapList(dynamic value) {
    if (value is List) {
      return value.map((item) {
        if (item is Map) {
          return {
            'url': item['url']?.toString() ?? '',
            'name': item['name']?.toString() ?? 'Untitled File',
          };
        }
        return <String, String>{};
      }).where((element) => element.isNotEmpty).toList();
    }
    return const [];
  }

  factory Task.fromJson(Map<String, dynamic> json) {
    final client = json['client'] is Map ? json['client'] : null;
    final guestInfo = json['guestInfo'] is Map ? json['guestInfo'] : null;
    final requestedStudent =
        json['requestedStudent'] is Map ? json['requestedStudent'] : null;
    final student = (json['student'] ?? json['assignedTo']) is Map
        ? (json['student'] ?? json['assignedTo'])
        : null;
    final submission = json['submission'] is Map ? json['submission'] : null;

    return Task(
      id: (json['_id'] ?? json['id'] ?? '').toString(),
      title: json['title']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      clientId: client?['_id']?.toString(),
      clientName: client?['name']?.toString(),
      budget: _toDoubleOrNull(json['budget']),
      status: json['status']?.toString() ?? 'open',
      location: json['location']?.toString(),
      domain: json['domain']?.toString(),
      company: json['company']?.toString() ?? client?['company']?.toString(),
      requiredSkills: _toStringList(json['requiredSkills']),
      deadline: json['deadline']?.toString(),
      attachments: _toStringList(json['attachments']),
      attachmentNames: _toStringList(json['attachmentNames']),
      requestedStudentId: requestedStudent?['_id']?.toString(),
      requestedStudentName: requestedStudent?['name']?.toString(),
      assignedStudentId: student?['_id']?.toString(),
      assignedStudentName: student?['name']?.toString(),
      clientTermsAccepted: _toBool(json['clientAgreedToTerms']),
      studentTermsAccepted: _toBool(json['studentAgreedToTerms']),
      
      // UPDATED MAPPING: Handle multiple files
      submissionFiles: _toFileMapList(submission?['files']),
      
      submissionNotes: submission?['notes']?.toString(),
      submittedAt: submission?['submittedAt']?.toString(),
      submissionApproved: _toBool(submission?['approved']),
      clientApproved: _toBool(submission?['approved']),
      rating: int.tryParse(json['rating']?.toString() ?? '0') ?? 0,
      score: int.tryParse(json['score']?.toString() ?? '0') ?? 0,
      feedback: json['feedback']?.toString() ?? '',
      isGuestTask: _toBool(json['isGuestTask']),
      guestName: guestInfo?['name']?.toString(),
      guestMobile: guestInfo?['mobile']?.toString(),
      guestEmail: guestInfo?['email']?.toString(),
      clientCanViewSubmission:
          _toBool(json['clientCanViewSubmission'], fallback: true),
      clientCanDownload: _toBool(json['clientCanDownload'], fallback: false),
      adminReceivedPayment: _toBool(json['adminReceivedPayment']),
      adminPaidStudent: _toBool(json['adminPaidStudent']),
      budgetFinalized: _toBool(json['budgetFinalized']),
      modificationNotes: json['modificationNotes']?.toString() ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'title': title,
      'description': description,
      'status': status,
      'isGuestTask': isGuestTask,
      'clientCanViewSubmission': clientCanViewSubmission,
      'clientCanDownload': clientCanDownload,
      'adminReceivedPayment': adminReceivedPayment,
      'adminPaidStudent': adminPaidStudent,
      'budgetFinalized': budgetFinalized,
      'budget': budget,
      'modificationNotes': modificationNotes,
      // Note: We don't usually send full submission objects back in toJson for task updates,
      // but if needed, you would export submissionFiles here.
    };
  }

  Task copyWith({
    String? id,
    String? title,
    String? description,
    String? status,
    bool? clientCanViewSubmission,
    bool? clientCanDownload,
    bool? adminReceivedPayment,
    bool? adminPaidStudent,
    bool? budgetFinalized,
    double? budget,
    String? modificationNotes,
    List<Map<String, String>>? submissionFiles,
  }) {
    return Task(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      clientId: clientId,
      clientName: clientName,
      budget: budget ?? this.budget,
      status: status ?? this.status,
      location: location,
      domain: domain,
      company: company,
      requiredSkills: requiredSkills,
      deadline: deadline,
      attachments: attachments,
      attachmentNames: attachmentNames,
      requestedStudentId: requestedStudentId,
      requestedStudentName: requestedStudentName,
      assignedStudentId: assignedStudentId,
      assignedStudentName: assignedStudentName,
      clientTermsAccepted: clientTermsAccepted,
      studentTermsAccepted: studentTermsAccepted,
      submissionFiles: submissionFiles ?? this.submissionFiles,
      submissionNotes: submissionNotes,
      submittedAt: submittedAt,
      submissionApproved: submissionApproved,
      clientApproved: clientApproved,
      rating: rating,
      score: score,
      feedback: feedback,
      isGuestTask: isGuestTask,
      guestName: guestName,
      guestMobile: guestMobile,
      guestEmail: guestEmail,
      clientCanViewSubmission:
          clientCanViewSubmission ?? this.clientCanViewSubmission,
      clientCanDownload: clientCanDownload ?? this.clientCanDownload,
      adminReceivedPayment: adminReceivedPayment ?? this.adminReceivedPayment,
      adminPaidStudent: adminPaidStudent ?? this.adminPaidStudent,
      budgetFinalized: budgetFinalized ?? this.budgetFinalized,
      modificationNotes: modificationNotes ?? this.modificationNotes,
    );
  }

  bool get isOpen => status == 'open';
  bool get isAssigned => status == 'assigned';
  bool get isUnderReview => status == 'under_review';
  bool get isCompleted => status == 'completed';
  
  // UPDATED: Now checks if the list contains files
  bool get hasSubmission => submissionFiles.isNotEmpty;
}